draft-specifications/CapTP Specification.md

```diff
@@ -0,0 +1,978 @@
+            ['fetch                  ; Argument 1: Symbol "fetch"
+             swiss-number]           ; Argument 2: Binary Data
+            3                        ; Answer position
+            <desc:import-object 5>>  ; object exported by us at position 5 should provide the answer
+```
+
+#### `deposit-gift` Method
+
+The deposit gift method is used in conjunction with sending a [Third Party
+Handoff](#third-party-handoffs). This method is used to deposit a gift which has
+been sent to the bootstrap object. It has two arguments:
+
+1.  A gift ID that is positive integer.
+2.  A `desc:import-object` or `desc:import-promise` which has been exported
+    within the given CapTP session.
+
+Here is an example of how to use this method:
+
+```text
+<op:deliver-only <desc:export 0>                               ; Remote bootstrap object
+                  ['deposit-gift                               ; Argument 1: Symbol "deposit-gift"
+                  gift-id                                      ; Argument 2: Positive integer or 0
+                  <desc:import-object | desc:import-promise>)> ; Argument 3
+```
+
+#### `withdraw-gift` Method
+
+This method is used to send the [`desc:handoff-receive`](#desc-handoff-receive)
+in order to receive a gift. It has one arguments:
+
+- The `desc:handoff-receive`
+
+This should have been sent with the `op:deliver` operation, the response the
+bootstrap object should give is the gift which was (or will be) deposited.
+
+Here is an example of how to use this method:
+
+```text
+<op:deliver <desc:export 0>           ; Remote bootstrap object
+            [withdraw-gift            ; Argument 1: Symbol "withdraw-gift"
+             <desc:handoff-receive>]  ; Argument 2: desc:handoff-receive
+            1                         ; Positive integer
+            <desc:import-object 3>>   ; The object exported (by us) at position 3, should receive the gift. 
+``` 
+
+## [`op:deliver-only`](#op-deliver-only)
+
+This operation delivers a message to an object and does not expect any result
+from the receiving object.
+
+The `op:deliver-only` message is:
+
+```text
+<op:deliver-only [to-desc   ; desc:export
+                  args]>    ; Sequence
+```
+
+`to-desc` is a `desc:export` descriptor which corresponds to the object the
+message is being sent to.
+
+`args` is a sequence which are the arguments to invoke the object with.
+
+Since this message has no result, it should be delivered to the object but
+nothing further is required.
+
+## [`op:deliver`](#op-deliver)
+
+This operation delivers a message to an object with the expectation of a return
+value which should be installed at the location specified in `answer-pos`. The
+`op:deliver` message is:
+
+``` text
+<op:deliver [to-desc            ; desc:export
+             args               ; sequence
+             answer-pos         ; positive integer | false
+             resolve-me-desc]>  ; desc:import-object | desc:import-promise
```

Comment:
resolve-me-desc here and in op:bootstrap seem to be redundant functionality of what op:listen does. See #15 for why.

Reply:
I actually hadn't seen #15 before so thank you for linking it. I think the concept of dropping the `resolve-me-desc` and then simplifying the `op:deliver` and `op:deliver-only` messages into a single `op:deliver` message and then relying on `op:listen`more is really interesting. I discussed it briefly with Christine and she thinks it's an interesting idea too.

I think this for now, this would be better addressed as a follow up PR if the group generally think this is a good idea (I think my initial impression is that I agree, this might be a nicer solution) we go ahead and make the change later.